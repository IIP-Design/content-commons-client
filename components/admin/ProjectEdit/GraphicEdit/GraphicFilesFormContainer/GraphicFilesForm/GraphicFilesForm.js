import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { useMutation } from '@apollo/react-hooks';
import { Confirm, Form, Input, Loader } from 'semantic-ui-react';
import ConfirmModalContent from 'components/admin/ConfirmModalContent/ConfirmModalContent';
import Filename from 'components/admin/Filename/Filename';
import FileRemoveReplaceButtonGroup from 'components/admin/FileRemoveReplaceButtonGroup/FileRemoveReplaceButtonGroup';
import FileUploadProgressBar from 'components/admin/FileUploadProgressBar/FileUploadProgressBar';
import FormikAutoSave from 'components/admin/FormikAutoSave/FormikAutoSave';
import GraphicStyleDropdown from 'components/admin/dropdowns/GraphicStyleDropdown/GraphicStyleDropdown';
import LanguageDropdown from 'components/admin/dropdowns/LanguageDropdown/LanguageDropdown';
import SocialPlatformDropdown from 'components/admin/dropdowns/SocialPlatformDropdown/SocialPlatformDropdown';
import VisuallyHidden from 'components/VisuallyHidden/VisuallyHidden';
import { UPDATE_GRAPHIC_PROJECT_MUTATION } from 'lib/graphql/queries/graphic';
import { formatBytes, getCount } from 'lib/utils';
import './GraphicFilesForm.scss';

const GraphicFilesForm = props => {
  const {
    errors,
    files,
    projectId,
    save,
    setFieldValue,
    setFieldTouched,
    touched,
    values,
  } = props;

  const [fileIdToDelete, setFileIdToDelete] = useState( '' );
  const [deleteConfirmOpen, setDeleteConfirmOpen] = useState( false );
  const [updateGraphicProject] = useMutation( UPDATE_GRAPHIC_PROJECT_MUTATION );

  const handleReset = () => {
    setDeleteConfirmOpen( false );
    setFileIdToDelete( '' );
  };

  const handleDelete = async id => {
    await updateGraphicProject( {
      variables: {
        data: {
          type: 'SOCIAL_MEDIA',
          images: {
            'delete': {
              id,
            },
          },
        },
        where: {
          id: projectId,
        },
      },
    } )
      .then( handleReset )
      .catch( err => console.dir( err ) );
  };

  const handleOnChange = ( e, { name, value } ) => {
    setFieldValue( name, value );
    setFieldTouched( name, true, false );
  };

  const isTouched = ( id, field ) => touched?.[id] && touched[id][field];

  // eslint-disable-next-line no-extra-parens
  const showErrorMsg = ( id, field ) => ( isTouched( id, field ) ? errors?.[id] && errors[id][field] : '' );
  const renderThumbnail = ( image, filename ) => {
    const imgSrc = image.signedUrl || image?.input?.dataUrl;
    const imgAlt = image.alt || filename;

    if ( imgSrc ) {
      return (
        <img
          src={ imgSrc }
          alt={ imgAlt }
          className="thumbnail"
        />
      );
    }

    return null;
  };

  const uploadInProgress = false; // temp

  if ( !getCount( files ) ) {
    return (
      <div className="graphic-project-graphic-files">
        <p className="no-files">
          { projectId
            ? 'Please upload at least one graphic file.'
            : 'No files to upload' }
        </p>
      </div>
    );
  }

  const hasFormBeenUpdated = !!Object.keys( touched ).length;

  return (
    <div className="graphic-project-graphic-files">
      { projectId && hasFormBeenUpdated && <FormikAutoSave save={ save } /> }

      <Confirm
        className="delete"
        open={ deleteConfirmOpen }
        content={ (
          <ConfirmModalContent
            className="delete_confirm"
            headline="Are you sure you want to delete this graphic?"
          >
            <p>This graphic will be permanently removed from the Content Commons and any other projects or collections it appears on.</p>
          </ConfirmModalContent>
        ) }
        onCancel={ handleReset }
        onConfirm={ () => handleDelete( fileIdToDelete ) }
        cancelButton="No, take me back"
        confirmButton="Yes, delete forever"
      />

      <Form className="form-fields">
        { /* We may have to pull the fieldset into a separate component and memoize it
        as all forms are rendered on each keystroke when updating the title
        Check performance before optimizing */ }
        { files.map( file => {
          const { id, filename, filesize, input, language } = file;
          const _filename = projectId ? filename : input?.name;
          const _filesize = projectId ? filesize : input?.size;

          const value = projectId
            ? values[id]
            : {
              title: input.name || '',
              language,
              style: file.style,
              social: file.social,
            };

          return (
            <fieldset
              key={ id }
              className={ `graphic-file-${id}` }
              name={ _filename }
            >
              <VisuallyHidden el="legend">
                { `edit fields for ${_filename}` }
              </VisuallyHidden>

              <div
                className={ `image-wrapper ${projectId ? 'available' : 'unavailable'}` }
              >
                { renderThumbnail( file, value?.title ) }

                <FileRemoveReplaceButtonGroup
                  onRemove={ () => {
                    setDeleteConfirmOpen( true );
                    setFileIdToDelete( id );
                  } }
                  icon="trash"
                />

                { uploadInProgress && <Loader active size="small" /> }

                <div className="meta-wrap">
                  <div className="meta">
                    <span className="filename">
                      <Filename
                        children={ _filename }
                        filenameLength={ 30 }
                        numCharsBeforeBreak={ 20 }
                        numCharsAfterBreak={ 10 }
                      />
                    </span>

                    <span className="filesize">
                      { formatBytes( _filesize || 0, 1 ) }
                    </span>
                  </div>
                </div>
              </div>

              { uploadInProgress // more conditions needed?
                && (
                  <div className="progress-bar">
                    <button
                      className="cancel-upload"
                      onClick={ () => { console.log( 'cancel' ); } }
                      type="button"
                    >
                      Cancel
                    </button>
                    <FileUploadProgressBar
                      filesToUpload={ [] } // temp
                      barSize="tiny"
                      onComplete={ () => {} } // temp
                      showPercent
                    />
                  </div>
                ) }

              <div className="field">
                <Form.Field
                  id={ `title-${id}` }
                  name={ `${id}.title` }
                  control={ Input }
                  label="Graphic Title"
                  value={ value?.title || '' }
                  onChange={ handleOnChange }
                  className={ language.textDirection }
                  disabled={ !projectId }
                />
              </div>

              <div className="form-group two-col">
                <div className="language-field">
                  <LanguageDropdown
                    id={ `language-${id}` }
                    name={ `${id}.language` }
                    className="language"
                    label="Language"
                    value={ value?.language || '' }
                    error={ isTouched( id, 'language' ) && !value.language }
                    onChange={ handleOnChange }
                    disabled={ !projectId }
                    required
                  />
                  <p className="error-message">
                    { showErrorMsg( id, 'language' ) }
                  </p>
                </div>

                <div className="graphic-style-field">
                  <GraphicStyleDropdown
                    id={ `graphic-style-${id}` }
                    name={ `${id}.style` }
                    className="graphic-style"
                    omit={ ['Clean'] }
                    label="Style"
                    value={ value?.style || '' }
                    error={ isTouched( id, 'style' ) && !value.style }
                    onChange={ handleOnChange }
                    disabled={ !projectId }
                    required
                  />
                  <p className="error-message">
                    { showErrorMsg( id, 'style' ) }
                  </p>
                </div>
              </div>

              <div className="social-platform-field">
                <SocialPlatformDropdown
                  id={ `social-platform-${id}` }
                  name={ `${id}.social` }
                  className="social-platform"
                  label="Platform"
                  value={ value?.social || '' }
                  error={ isTouched( id, 'social' ) && !value.social }
                  onChange={ handleOnChange }
                  disabled={ !projectId }
                  multiple
                  search
                  closeOnBlur
                  closeOnChange
                  required
                />
                <p className="error-message">
                  { showErrorMsg( id, 'social' ) }
                </p>
              </div>
            </fieldset>
          );
        } ) }
      </Form>
    </div>
  );
};

GraphicFilesForm.propTypes = {
  errors: PropTypes.object,
  files: PropTypes.array,
  projectId: PropTypes.string,
  save: PropTypes.func,
  setFieldValue: PropTypes.func,
  setFieldTouched: PropTypes.func,
  touched: PropTypes.object,
  values: PropTypes.object,
};

export default GraphicFilesForm;
