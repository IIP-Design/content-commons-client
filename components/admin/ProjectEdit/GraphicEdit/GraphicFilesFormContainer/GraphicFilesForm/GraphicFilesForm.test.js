import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import wait from 'waait';
import { MockedProvider } from '@apollo/react-testing';
import { formatBytes, truncateAndReplaceStr } from 'lib/utils';
import { mocks, props } from './mocks';
import GraphicFilesForm from './GraphicFilesForm';

jest.mock(
  'components/admin/ConfirmModalContent/ConfirmModalContent',
  () => function ConfirmModalContent() { return ''; },
);

jest.mock(
  'components/admin/FileRemoveReplaceButtonGroup/FileRemoveReplaceButtonGroup',
  () => function FileRemoveReplaceButtonGroup() { return ''; },
);

jest.mock(
  'components/admin/FileUploadProgressBar/FileUploadProgressBar',
  () => function FileUploadProgressBar() { return ''; },
);

jest.mock(
  'components/admin/FormikAutoSave/FormikAutoSave',
  () => function FormikAutoSave() { return ''; },
);

jest.mock(
  'components/admin/dropdowns/GraphicStyleDropdown/GraphicStyleDropdown',
  () => function GraphicStyleDropdown() { return ''; },
);

jest.mock(
  'components/admin/dropdowns/LanguageDropdown/LanguageDropdown',
  () => function LanguageDropdown() { return ''; },
);

jest.mock(
  'components/admin/dropdowns/SocialPlatformDropdown/SocialPlatformDropdown',
  () => function SocialPlatformDropdown() { return ''; },
);

const suppressActWarning = consoleError => {
  const actMsg = 'Warning: An update to %s inside a test was not wrapped in act';

  jest.spyOn( console, 'error' ).mockImplementation( ( ...args ) => {
    if ( !args[0].includes( actMsg ) ) {
      consoleError( ...args );
    }
  } );
};

describe( '<GraphicFilesForm />', () => {
  const consoleError = console.error;

  beforeAll( () => suppressActWarning( consoleError ) );
  afterAll( () => {
    console.error = consoleError;
  } );

  let Component;
  let wrapper;
  let graphicsForm;
  let fieldsets;

  beforeEach( () => {
    Component = (
      <MockedProvider mocks={ mocks } addTypename>
        <GraphicFilesForm { ...props } />
      </MockedProvider>
    );

    wrapper = mount( Component );
    graphicsForm = wrapper.find( 'GraphicFilesForm' );
    fieldsets = graphicsForm.find( 'fieldset' );
  } );

  it( 'renders without crashing', () => {
    expect( graphicsForm.exists() ).toEqual( true );
  } );

  it( 'renders with the correct className value', () => {
    const containingDiv = graphicsForm.find( '.graphic-project-graphic-files' );

    expect( containingDiv.exists() ).toEqual( true );
  } );

  it( 'renders the Form with the correct className value', () => {
    const semanticForm = graphicsForm.find( 'Form' );

    expect( semanticForm.exists() ).toEqual( true );
    expect( semanticForm.hasClass( 'form-fields' ) ).toEqual( true );
  } );

  it( 'renders the Confirm modal', () => {
    const confirm = graphicsForm.find( 'Confirm' );
    const content = mount( confirm.prop( 'content' ) );
    const contentMsg = mount( content.prop( 'children' ) );

    expect( confirm.exists() ).toEqual( true );
    expect( confirm.hasClass( 'delete' ) ).toEqual( true );
    expect( confirm.prop( 'open' ) ).toEqual( false );
    expect( content.prop( 'className' ) ).toEqual( 'delete_confirm' );
    expect( content.prop( 'headline' ) )
      .toEqual( 'Are you sure you want to delete this graphic?' );
    expect( contentMsg.name() ).toEqual( 'p' );
    expect( contentMsg.text() )
      .toEqual( 'This graphic will be permanently removed from the Content Commons and any other projects or collections it appears on.' );
  } );

  it( 'clicking the trash icon opens the Confirm modal', () => {
    const replaceBtn = fieldsets.first().find( 'FileRemoveReplaceButtonGroup' );

    // open the modal
    replaceBtn.prop( 'onRemove' )();
    wrapper.update();
    const confirm = wrapper.find( 'Confirm' );

    expect( confirm.prop( 'open' ) ).toEqual( true );
  } );

  it( 'clicking the Confirm modal "No, take me back" button closes the modal', () => {
    const replaceBtn = fieldsets.first().find( 'FileRemoveReplaceButtonGroup' );

    // open the modal
    replaceBtn.prop( 'onRemove' )();
    wrapper.update();
    const confirm = () => wrapper.find( 'Confirm' );

    expect( confirm().prop( 'open' ) ).toEqual( true );

    // close the modal
    confirm().prop( 'onCancel' )();
    wrapper.update();

    expect( confirm().prop( 'open' ) ).toEqual( false );
  } );

  it( 'clicking the Confirm modal "Yes, delete forever" button', () => {
    const replaceBtn = fieldsets.first().find( 'FileRemoveReplaceButtonGroup' );

    // open the modal
    replaceBtn.prop( 'onRemove' )();
    wrapper.update();
    const confirm = () => wrapper.find( 'Confirm' );

    expect( confirm().prop( 'open' ) ).toEqual( true );

    // confirm deletion
    const deleteConfirmTest = async done => {
      confirm().prop( 'onConfirm' )();
      await wait( 4 ); // wait for mutation to resolve
      wrapper.update();

      // handleReset is called upon successful deletion
      expect( confirm().prop( 'open' ) ).toEqual( false );
      done();
    };

    deleteConfirmTest();
  } );

  it( 'renders the correct fieldsets', () => {
    expect( fieldsets.length ).toEqual( props.files.length );
    expect( toJSON( fieldsets ) ).toMatchSnapshot();
  } );

  it( 'changing the title field value calls handleOnChange', () => {
    const { id } = props.files[0];
    const titleField = graphicsForm.find( `Input#title-${id}` );
    const e = {};
    const data = {
      name: `${id}.title`,
      value: 'new title',
    };

    titleField.prop( 'onChange' )( e, data );
    expect( props.setFieldValue )
      .toHaveBeenCalledWith( data.name, data.value );
    expect( props.setFieldTouched )
      .toHaveBeenCalledWith( data.name, true, false );
  } );

  it( 'changing the language value calls handleOnChange', () => {
    const { id } = props.files[0];
    const language = graphicsForm.find( `#graphic-style-${id}` );
    const e = {};
    const data = {
      name: `${id}.language`,
      value: '8888888',
    };

    language.prop( 'onChange' )( e, data );
    expect( props.setFieldValue )
      .toHaveBeenCalledWith( data.name, data.value );
    expect( props.setFieldTouched )
      .toHaveBeenCalledWith( data.name, true, false );
  } );

  it( 'changing the graphic style value calls handleOnChange', () => {
    const { id } = props.files[0];
    const style = graphicsForm.find( `#graphic-style-${id}` );
    const e = {};
    const data = {
      name: `${id}.style`,
      value: '8888888',
    };

    style.prop( 'onChange' )( e, data );
    expect( props.setFieldValue )
      .toHaveBeenCalledWith( data.name, data.value );
    expect( props.setFieldTouched )
      .toHaveBeenCalledWith( data.name, true, false );
  } );

  it( 'changing the social platform value calls handleOnChange', () => {
    const { id } = props.files[0];
    const social = graphicsForm.find( `#social-platform-${id}` );
    const e = {};
    const data = {
      name: `${id}.social`,
      value: '8888888',
    };

    social.prop( 'onChange' )( e, data );
    expect( props.setFieldValue )
      .toHaveBeenCalledWith( data.name, data.value );
    expect( props.setFieldTouched )
      .toHaveBeenCalledWith( data.name, true, false );
  } );
} );

describe( '<GraphicFilesForm />, when no files are received', () => {
  const consoleError = console.error;

  beforeAll( () => suppressActWarning( consoleError ) );
  afterAll( () => {
    console.error = consoleError;
  } );

  let Component;
  let wrapper;
  let graphicsForm;

  const newProps = {
    ...props,
    files: [],
  };

  beforeEach( () => {
    Component = (
      <MockedProvider mocks={ mocks } addTypename>
        <GraphicFilesForm { ...newProps } />
      </MockedProvider>
    );

    wrapper = mount( Component );
    graphicsForm = wrapper.find( 'GraphicFilesForm' );
  } );

  it( 'renders a No Files message', () => {
    const semanticForm = graphicsForm.find( 'Form' );
    const noFilesWrapper = graphicsForm.find( '.no-files' );
    const msg = 'No files to upload';

    expect( semanticForm.exists() ).toEqual( false );
    expect( noFilesWrapper.exists() ).toEqual( true );
    expect( noFilesWrapper.name() ).toEqual( 'p' );
    expect( noFilesWrapper.contains( msg ) ).toEqual( true );
  } );
} );

describe( '<GraphicFilesForm />, when a required field value is missing', () => {
  const consoleError = console.error;

  beforeAll( () => suppressActWarning( consoleError ) );
  afterAll( () => {
    console.error = consoleError;
  } );

  let Component;
  let wrapper;
  let graphicsForm;

  const fileId = props.files[0].id;

  const newProps = {
    ...props,
    errors: {
      [fileId]: {
        language: 'A language is required',
        style: 'A graphic style is required.',
        social: 'A social platform is required.',
      },
    },
    touched: {
      [fileId]: {
        language: true,
        style: true,
        social: true,
      },
    },
    values: {
      ...props.values,
      [fileId]: {
        ...props.values[fileId],
        language: null,
        style: null,
        social: null,
      },
    },
  };

  beforeEach( () => {
    Component = (
      <MockedProvider mocks={ mocks } addTypename>
        <GraphicFilesForm { ...newProps } />
      </MockedProvider>
    );

    wrapper = mount( Component );
    graphicsForm = wrapper.find( 'GraphicFilesForm' );
  } );

  it( 'renders FormikAutoSave since fields are touched', () => {
    const formikAutoSave = graphicsForm.find( 'FormikAutoSave' );

    expect( formikAutoSave.exists() ).toEqual( true );
    expect( typeof formikAutoSave.prop( 'save' ) ).toEqual( 'function' );
  } );

  it( 'renders an error message if there is no language value', () => {
    const languageWrappers = graphicsForm.find( '.language-field' );
    const language = graphicsForm.find( `#language-${fileId}` );
    const msg = newProps.errors[fileId].language;

    expect( languageWrappers.exists() ).toEqual( true );
    expect( language.prop( 'error' ) ).toEqual( true );
    expect( languageWrappers.first().contains( msg ) ).toEqual( true );
  } );

  it( 'renders an error message if there is no graphic style value', () => {
    const styleWrappers = graphicsForm.find( '.graphic-style-field' );
    const style = graphicsForm.find( `#graphic-style-${fileId}` );
    const msg = newProps.errors[fileId].style;

    expect( styleWrappers.exists() ).toEqual( true );
    expect( style.prop( 'error' ) ).toEqual( true );
    expect( styleWrappers.first().contains( msg ) ).toEqual( true );
  } );

  it( 'renders an error message if there is no social platform value', () => {
    const socialWrappers = graphicsForm.find( '.social-platform-field' );
    const social = graphicsForm.find( `#social-platform-${fileId}` );
    const msg = newProps.errors[fileId].social;

    expect( socialWrappers.exists() ).toEqual( true );
    expect( social.prop( 'error' ) ).toEqual( true );
    expect( socialWrappers.first().contains( msg ) ).toEqual( true );
  } );
} );

describe( '<GraphicFilesForm />, when a projectId does not exist', () => {
  const consoleError = console.error;

  beforeAll( () => suppressActWarning( consoleError ) );
  afterAll( () => {
    console.error = consoleError;
  } );

  let Component;
  let wrapper;
  let graphicsForm;
  let fieldsets;

  const newProps = {
    ...props,
    projectId: undefined,
    files: [
      {
        id: 'de71e989-5701-4d28-ad1d-9fb152016ab7',
        name: '4_3_Serious_FB.jpg',
        loaded: 0,
        input: {
          dataUrl: 'the-data-url',
          name: '4_3_Serious_FB.jpg',
          size: 433851,
          __typename: 'LocalInputFile',
        },
        language: 'ck2lzfx710hkq07206thus6pt',
        style: 'ck9h3kyb326ak0720wkbk01q6',
        social: ['ck9h3m9bl26bm0720rm69c60s', 'ck9h3meu626bw07201o36tapc'],
        __typename: 'LocalImageFile',
      },
      {
        id: '727a755c-8274-4df5-8417-f7720c06b6eb',
        name: '4_3_Serious_TW.jpg',
        loaded: 0,
        input: {
          dataUrl: 'the-data-url',
          name: '4_3_Serious_TW.jpg',
          size: 297343,
          __typename: 'LocalInputFile',
        },
        language: 'ck2lzfx710hkq07206thus6pt',
        style: 'ck9h3ka3o269y0720t7wzp5uq',
        social: ['ck9h3m3g626bd07201gh712vk'],
        __typename: 'LocalImageFile',
      },
    ],
  };

  beforeEach( () => {
    Component = (
      <MockedProvider mocks={ [] } addTypename>
        <GraphicFilesForm { ...newProps } />
      </MockedProvider>
    );

    wrapper = mount( Component );
    graphicsForm = wrapper.find( 'GraphicFilesForm' );
    fieldsets = graphicsForm.find( 'fieldset' );
  } );

  it( 'renders a disabled graphic title field', () => {
    const titleFields = graphicsForm.find( 'input[id^="title-"]' );

    titleFields.forEach( ( field, i ) => {
      const { id, input } = newProps.files[i];
      const imgWrapper = fieldsets.find( `.graphic-file-${id} > .image-wrapper` );

      expect( field.prop( 'name' ) ).toEqual( `${id}.title` );
      expect( field.prop( 'value' ) ).toEqual( input.name );
      expect( field.prop( 'disabled' ) ).toEqual( true );
      expect( imgWrapper.hasClass( 'unavailable' ) ).toEqual( true );
    } );
  } );

  it( 'renders a disabled language dropdown', () => {
    const langDropdowns = graphicsForm.find( '[id^="language-"]' );

    langDropdowns.forEach( ( field, i ) => {
      const { id, language } = newProps.files[i];
      const imgWrapper = fieldsets.find( `.graphic-file-${id} > .image-wrapper` );

      expect( field.prop( 'name' ) ).toEqual( `${id}.language` );
      expect( field.prop( 'value' ) ).toEqual( language );
      expect( field.prop( 'disabled' ) ).toEqual( true );
      expect( imgWrapper.hasClass( 'unavailable' ) ).toEqual( true );
    } );
  } );

  it( 'renders a disabled graphic style dropdown', () => {
    const styleFields = graphicsForm.find( '[id^="graphic-style-"]' );

    styleFields.forEach( ( field, i ) => {
      const { id, style } = newProps.files[i];
      const imgWrapper = fieldsets.find( `.graphic-file-${id} > .image-wrapper` );

      expect( field.prop( 'name' ) ).toEqual( `${id}.style` );
      expect( field.prop( 'value' ) ).toEqual( style );
      expect( field.prop( 'disabled' ) ).toEqual( true );
      expect( imgWrapper.hasClass( 'unavailable' ) ).toEqual( true );
    } );
  } );

  it( 'renders a disabled social platform dropdown', () => {
    const socialFields = graphicsForm.find( '[id^="social-platform-"]' );

    socialFields.forEach( ( field, i ) => {
      const { id, social } = newProps.files[i];
      const imgWrapper = fieldsets.find( `.graphic-file-${id} > .image-wrapper` );

      expect( field.prop( 'name' ) ).toEqual( `${id}.social` );
      expect( field.prop( 'value' ) ).toEqual( social );
      expect( field.prop( 'disabled' ) ).toEqual( true );
      expect( imgWrapper.hasClass( 'unavailable' ) ).toEqual( true );
    } );
  } );

  it( 'renders the filename correctly', () => {
    const filenames = graphicsForm.find( '.filename' );

    filenames.forEach( ( filename, i ) => {
      const inputName = newProps.files[i].input.name;

      expect( filename.contains( inputName ) ).toEqual( true );
    } );
  } );

  it( 'renders the filesize correctly', () => {
    const filesizes = graphicsForm.find( '.filesize' );

    filesizes.forEach( ( filesize, i ) => {
      const inputSize = formatBytes( newProps.files[i].input.size, 1 );

      expect( filesize.text() ).toEqual( inputSize );
    } );
  } );
} );

describe( '<GraphicFilesForm />, when there is a long file name', () => {
  const consoleError = console.error;

  beforeAll( () => suppressActWarning( consoleError ) );
  afterAll( () => {
    console.error = consoleError;
  } );

  let Component;
  let wrapper;
  let graphicsForm;
  let fieldsets;


  const newProps = {
    ...props,
    files: [
      {
        ...props.files[0],
        filename: '4_3_Serious_aieaue_kaienwiz_ke8akcua_aeicaie_eiamwyz_TW.jpg',
      },
    ],
  };

  beforeEach( () => {
    Component = (
      <MockedProvider mocks={ mocks } addTypename>
        <GraphicFilesForm { ...newProps } />
      </MockedProvider>
    );

    wrapper = mount( Component );
    graphicsForm = wrapper.find( 'GraphicFilesForm' );
    fieldsets = graphicsForm.find( 'fieldset' );
  } );

  it( 'renders a truncated filename with the full filename visually hidden', () => {
    fieldsets.forEach( ( fieldset, i ) => {
      const { filename } = newProps.files[i];
      const imgWrapper = fieldset.find( '.image-wrapper' );
      const btn = imgWrapper.find( '.filename.truncated' );
      const visuallyHidden = imgWrapper.find( '.hide-visually' );
      const displayName = filename?.length > 30
        ? truncateAndReplaceStr( filename, 20, 10 )
        : filename;

      expect( btn.prop( 'tooltip' ) ).toEqual( filename );
      expect( btn.text() ).toEqual( displayName );
      expect( visuallyHidden.text() ).toEqual( filename );
    } );
  } );
} );
