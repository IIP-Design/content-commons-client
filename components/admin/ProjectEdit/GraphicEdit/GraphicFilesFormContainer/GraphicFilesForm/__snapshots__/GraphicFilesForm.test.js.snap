// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GraphicFilesForm /> renders the correct fieldsets 1`] = `
Array [
  <fieldset
    className="graphic-file-cka5di41430i10720v06c52ic"
    key="cka5di41430i10720v06c52ic"
    name="4_3_Serious_TW.jpg"
  >
    <VisuallyHidden
      el="legend"
    >
      <legend
        className="hide-visually"
      >
        edit fields for 4_3_Serious_TW.jpg
      </legend>
    </VisuallyHidden>
    <div
      className="image-wrapper available"
    >
      <img
        alt="4_3_Serious_TW.jpg"
        className="thumbnail"
        src="https://amgov-publisher-dev.s3.amazonaws.com/2020/04/commons.america.gov_ck9laaua62c2o0720577s3jto/4_3_Serious_TW.jpg?AWSAccessKeyId=someaccesskey&Expires=1589466970&Signature=thesignature"
      />
      <FileRemoveReplaceButtonGroup
        icon="trash"
        onRemove={[Function]}
      />
      <div
        className="meta-wrap"
      >
        <div
          className="meta"
        >
          <span
            className="filename"
          >
            <Filename
              filenameLength={30}
              numCharsAfterBreak={10}
              numCharsBeforeBreak={20}
            >
              <div
                className="filename"
              >
                <span>
                  4_3_Serious_TW.jpg
                </span>
              </div>
            </Filename>
          </span>
          <span
            className="filesize"
          >
            290.4  KB
          </span>
        </div>
      </div>
    </div>
    <div
      className="field"
    >
      <FormField
        className="LTR"
        control={[Function]}
        disabled={false}
        id="title-cka5di41430i10720v06c52ic"
        label="Graphic Title"
        name="cka5di41430i10720v06c52ic.title"
        onChange={[Function]}
        value="4_3_Serious_TW.jpg"
      >
        <div
          className="field LTR"
        >
          <label
            htmlFor="title-cka5di41430i10720v06c52ic"
          >
            Graphic Title
          </label>
          <Input
            disabled={false}
            id="title-cka5di41430i10720v06c52ic"
            name="cka5di41430i10720v06c52ic.title"
            onChange={[Function]}
            type="text"
            value="4_3_Serious_TW.jpg"
          >
            <div
              className="ui input"
            >
              <input
                disabled={false}
                id="title-cka5di41430i10720v06c52ic"
                name="cka5di41430i10720v06c52ic.title"
                onChange={[Function]}
                type="text"
                value="4_3_Serious_TW.jpg"
              />
            </div>
          </Input>
        </div>
      </FormField>
    </div>
    <div
      className="form-group two-col"
    >
      <div
        className="language-field"
      >
        <LanguageDropdown
          className="language"
          disabled={false}
          id="language-cka5di41430i10720v06c52ic"
          label="Language"
          name="cka5di41430i10720v06c52ic.language"
          onChange={[Function]}
          required={true}
          value="ck2lzfx710hkq07206thus6pt"
        />
        <p
          className="error-message"
        />
      </div>
      <div
        className="graphic-style-field"
      >
        <GraphicStyleDropdown
          className="graphic-style"
          disabled={false}
          id="graphic-style-cka5di41430i10720v06c52ic"
          label="Style"
          name="cka5di41430i10720v06c52ic.style"
          omit={
            Array [
              "Clean",
            ]
          }
          onChange={[Function]}
          required={true}
          value="ck9h3koe426aa0720y421wmk3"
        />
        <p
          className="error-message"
        />
      </div>
    </div>
    <div
      className="social-platform-field"
    >
      <SocialPlatformDropdown
        className="social-platform"
        closeOnBlur={true}
        closeOnChange={true}
        disabled={false}
        id="social-platform-cka5di41430i10720v06c52ic"
        label="Platform"
        multiple={true}
        name="cka5di41430i10720v06c52ic.social"
        onChange={[Function]}
        required={true}
        search={true}
        value={
          Array [
            "ck9h3m3g626bd07201gh712vk",
          ]
        }
      />
      <p
        className="error-message"
      />
    </div>
  </fieldset>,
  <fieldset
    className="graphic-file-cka5di48r30ig072003dqjffb"
    key="cka5di48r30ig072003dqjffb"
    name="4_3_Serious_FB.jpg"
  >
    <VisuallyHidden
      el="legend"
    >
      <legend
        className="hide-visually"
      >
        edit fields for 4_3_Serious_FB.jpg
      </legend>
    </VisuallyHidden>
    <div
      className="image-wrapper available"
    >
      <img
        alt="4_3_Serious_FB.jpg"
        className="thumbnail"
        src="https://amgov-publisher-dev.s3.amazonaws.com/2020/04/commons.america.gov_ck9laaua62c2o0720577s3jto/4_3_Serious_FB.jpg?AWSAccessKeyId=someaccesskey&Expires=1589466970&Signature=thesignature"
      />
      <FileRemoveReplaceButtonGroup
        icon="trash"
        onRemove={[Function]}
      />
      <div
        className="meta-wrap"
      >
        <div
          className="meta"
        >
          <span
            className="filename"
          >
            <Filename
              filenameLength={30}
              numCharsAfterBreak={10}
              numCharsBeforeBreak={20}
            >
              <div
                className="filename"
              >
                <span>
                  4_3_Serious_FB.jpg
                </span>
              </div>
            </Filename>
          </span>
          <span
            className="filesize"
          >
            423.7  KB
          </span>
        </div>
      </div>
    </div>
    <div
      className="field"
    >
      <FormField
        className="LTR"
        control={[Function]}
        disabled={false}
        id="title-cka5di48r30ig072003dqjffb"
        label="Graphic Title"
        name="cka5di48r30ig072003dqjffb.title"
        onChange={[Function]}
        value="4_3_Serious_FB.jpg"
      >
        <div
          className="field LTR"
        >
          <label
            htmlFor="title-cka5di48r30ig072003dqjffb"
          >
            Graphic Title
          </label>
          <Input
            disabled={false}
            id="title-cka5di48r30ig072003dqjffb"
            name="cka5di48r30ig072003dqjffb.title"
            onChange={[Function]}
            type="text"
            value="4_3_Serious_FB.jpg"
          >
            <div
              className="ui input"
            >
              <input
                disabled={false}
                id="title-cka5di48r30ig072003dqjffb"
                name="cka5di48r30ig072003dqjffb.title"
                onChange={[Function]}
                type="text"
                value="4_3_Serious_FB.jpg"
              />
            </div>
          </Input>
        </div>
      </FormField>
    </div>
    <div
      className="form-group two-col"
    >
      <div
        className="language-field"
      >
        <LanguageDropdown
          className="language"
          disabled={false}
          id="language-cka5di48r30ig072003dqjffb"
          label="Language"
          name="cka5di48r30ig072003dqjffb.language"
          onChange={[Function]}
          required={true}
          value="ck2lzfx710hkq07206thus6pt"
        />
        <p
          className="error-message"
        />
      </div>
      <div
        className="graphic-style-field"
      >
        <GraphicStyleDropdown
          className="graphic-style"
          disabled={false}
          id="graphic-style-cka5di48r30ig072003dqjffb"
          label="Style"
          name="cka5di48r30ig072003dqjffb.style"
          omit={
            Array [
              "Clean",
            ]
          }
          onChange={[Function]}
          required={true}
          value="ck9h3kyb326ak0720wkbk01q6"
        />
        <p
          className="error-message"
        />
      </div>
    </div>
    <div
      className="social-platform-field"
    >
      <SocialPlatformDropdown
        className="social-platform"
        closeOnBlur={true}
        closeOnChange={true}
        disabled={false}
        id="social-platform-cka5di48r30ig072003dqjffb"
        label="Platform"
        multiple={true}
        name="cka5di48r30ig072003dqjffb.social"
        onChange={[Function]}
        required={true}
        search={true}
        value={
          Array [
            "ck9h3m9bl26bm0720rm69c60s",
          ]
        }
      />
      <p
        className="error-message"
      />
    </div>
  </fieldset>,
]
`;
